### PR TITLE
chore(gatsby): use lazy iterable instead of fully realized arrays for detecting stale nodes

### PR DESCRIPTION
**Pull request by pieh**
## Description

TL/DR: for now this is just swapping from array to iterable, but it doesn't actually solve potential high memory usage/OOM during this step - see explanation below

**UPDATE**:
There is a way to GC things actually in sync scenarios so while everything below is true today, there will be changes coming to `iterateNodes()` internals to allow for things to GC.

This alone in practice doesn't have an effect on memory usage due to function chain here being fully sync and us using lmdb's cache setting. I'm still researching it, but it seems like LMDB cache is using `WeakRef` and object that are wrapped with it CAN'T be garbage collected within same "tick" (?). Per [WeakRef spec](https://github.com/tc39/proposal-weakrefs/blob/99c96cfbaf919f8405d329e76df166e7aab34d7a/history/spec.md) creating or dereferencing WeakRef should perform `KeepDuringJob` operation

> When the abstract operation KeepDuringJob is called with a target object reference, it adds the target to an identity Set that will point strongly at the target until the end of the current Job.

Currently it's not clear to me wether `Job` in spec is same as "tick" or not, but Node.js implementation seem to rely on microtasks to clear it and allow objects to be GCed ( https://github.com/nodejs/node/blob/fd44fe5794e42c9cfbcadcba64a3805e0f6ce056/deps/v8/src/execution/microtask-queue.cc#L127 )

Note that I did try `expirer: false` setting to disable LRU part of lmdb cache and only rely on WeakRefs ( https://github.com/kriszyp/weak-lru-cache#expirer ), but given that WeakRef seems to be main culprit it didn't change anything in reality (at least for this step)

Possible solutions that I can currently think of:
 - disable `cache` on nodes datastore - this does allow for nodes that were already iterated on to be GCed, but downside is that every `getNode` call would get new separate object (`getNode(x) === getNode(x)` would no longer be true) and also could mean perf degration in cases when we/plugins call `getNode` for some id multiple times
 - consider `async` alternative for sync `GatsbyIterable` operations that would be able to do `process.nextTick` (or one of those) every X iterations to allow weakrefs to no longer by referenced by `KeepDuringJob` and ultimately be GCed - this looks like can of worms and I don't know the impact of it currently (good part is that we don't expose `iterateNodes` yet, so changing it to async might not be a breaking change at a first glance, but defenitely would require a lot of internal changes).

## Related Issues

[sc-44731]

### CODE REVIEW

**Comment by j0sh77** [comment]
This makes sense to me. Even though it's not (currently) a functionality change, it allows us to easily swap out `GatsbyIterable` and `GatsbyNode` implementations later on!

**Comment by j0sh77** [comment]
What do you intend with this PR going forward? I see you put this up as a draft, so I'm just curious what next steps are.

**Comment by pieh** [comment]
@j0sh77 I will be testing out wether usage of newly exposed `clearKeptObjects` in `lmdb-js` ( https://github.com/DoctorEvidence/lmdb-js/commit/82fac029d210e6a936bae0fcc86e59a95016ae69 ) will allow for this to actuall work regarding garbage collection.

As rule of thumb we do want to change any full arrays of nodes to use iterables, so that nodes already processed can be garbage collected. As of right now if we do `iterateNodes()` in a sync function it's pretty much same as using arrays because nodes can't be GCed.

If it turns out that exposed function doesn't work for us - we will have to pivot and add something like `await new Promise(r => setImmediate(r))` every 100 iterations (value to be determined) and change code paths to be async

**Comment by pieh** [comment]
I tested with https://github.com/DoctorEvidence/lmdb-js/commit/82fac029d210e6a936bae0fcc86e59a95016ae69 and https://github.com/gatsbyjs/gatsby/commit/755d94617b0f79674eb235cf0637d1b28ece3422 and this code works now if available memory is less than all nodes combined meaning we are on track and this won't need to be changed further (just will have to wait for other things to land to make an impact on memory retention)

File: packages/gatsby/src/utils/source-nodes.ts [hunk 1]
```diff
@@ -77,12 +81,13 @@ function getStaleNodes(state: IGatsbyState, nodes: Array<Node>): Array<Node> {
...
+function deleteStaleNodes(
+  state: IGatsbyState,
+  nodes: GatsbyIterable<IGatsbyNode>
+): void {
   const staleNodes = getStaleNodes(state, nodes)
 
-  if (staleNodes.length > 0) {
-    staleNodes.forEach(node => store.dispatch(deleteNode(node)))
-  }
+  staleNodes.forEach(node => store.dispatch(deleteNode(node)))
```
Comment (pieh) [inline packages/gatsby/src/utils/source-nodes.ts:90]: there is no `.length`on `GatsbyIterable`, but `.forEach` on iterables with 0 elements is largely no-op anyway

**Review by j0sh77 — APPROVED** [review]
Ship it
